### PR TITLE
Build the metatheory package and plc-agda executable

### PR DESCRIPTION
**Pull request by michaelpj**
We need a new version of Agda, which we get by adding it to our package
set. It's not on Stackage, so we end up adding it and a number of transitive
deps as `extra-deps`.

We then need to patch up the nixpkgs Agda infrastructure to use it, and
also get a new version of the Agda stdlib.

We also build the literate Agda output as HTML in the main
derivation, and add another derivation building the compiled Haskell
sources.

Finally, we add a derivation based on the `plc-agda` Cabal file that
uses the generated sources and builds the `plc-agda` executable.

FYI @disassembler / @angerman for interest. I don't know how much more Agda building we need to do as a company, but this seems to work okay.

### CODE REVIEW

**Comment by michaelpj** [comment]
And I tested it and the executable does work!

File: metatheory/plc-agda.nix [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+mkDerivation {
+  pname = "plc-agda";
+  version = "0.1.0.0";
+  src = ./.;
+  isLibrary = false;
+  isExecutable = true;
+  executableHaskellDepends = [
+    base bytestring cryptonite ieee754 language-plutus-core memory text
+  ];
+  homepage = "https://github.com/input-output-hk/plutus-metatheory";
```
Comment (jmchapman) [inline metatheory/plc-agda.nix]: shouldn't this be `homepage = "https://github.com/input-output-hk/plutus/metatheory"`?

File: metatheory/default.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  cleanSourceAgda = src: pkgs.lib.cleanSourceWith { filter = libFilter; src = (cleanSourceHaskell src); };
+in rec { 
+  plutus-formal = agda.mkDerivation (self: rec {
+    name = "plutus-formal";
+
+    buildDepends = [ AgdaStdlib ];
+
+    # We can't just add more flags, annoyingly, so we have to repeat some of the existing flags.
+    # Passing the html output flags gets us the literate output, and still checks the Agda, so this
+    # derivation can do double duty.
```
Comment (jmchapman) [inline metatheory/default.nix]: good idea to do checking and doc generation together

File: metatheory/default.nix [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    buildFlags = pkgs.lib.concatStringsSep " " (["--html" "--html-highlight=auto" ] ++ (map (x: "-i " + x) self.includeDirs));
```
Comment (jmchapman) [inline metatheory/default.nix]: does this mean it's easier to pull in the std-lib directly `-i` rather than using the agda library management stuff?

Comment (michaelpj) [reply metatheory/default.nix]: The nixpkgs agda builder doesn't support library stuff. I've opened an issue upstream - it's fairly unloved. Trying not to get sucked into writing a new one :grimacing:

File: metatheory/plc-agda.nix (hunk 1 above)
Comment (michaelpj) [reply metatheory/plc-agda.nix]: Ah yes, the cabal file is out of date, will fix.

**Pull request merged**